include/swift/AST/AutoDiff.h

```diff
@@ -0,0 +1,273 @@
+//===--- AutoDiff.h - Swift Differentiable Programming --------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2017 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+//
+//  This file defines AST support for the experimental differentiable
+//  programming feature.
+//
+//===----------------------------------------------------------------------===//
+
+
+#ifndef SWIFT_AST_AUTODIFF_H
+#define SWIFT_AST_AUTODIFF_H
+
+#include "ASTContext.h"
+#include "llvm/ADT/SmallBitVector.h"
+#include "swift/Basic/Range.h"
+
+namespace swift {
+
+/// An efficient index subset data structure, uniqued in `ASTContext`.
+/// Stores a bit vector representing set indices and a total capacity.
+class AutoDiffIndexSubset : public llvm::FoldingSetNode {
```

Comment:
I know this is used by autodiff, but it's a fairly general data structure. Should we name it something like `UniquedIndexSet`?

Reply:
Thanks for the suggestion! I think "uniqued" sounds good. The word "subset" here is to reflect the fact that the data structure stores a capacity and that it's used to denote a subset of indices. So, how about `UniquedIndexSubset`?

Reply:
Also, would you prefer it to be added to some other file?

Reply:
Actually I started to feel that "uniqued" could be redundant since other uniqued data structures don't have a "uniqued" prefix. How do you feel about just `IndexSubset`?

Reply:
Works for me. Not sure why it's a "subset" rather than a "set" ;)